plugins/inputs/windows_event_log/wineventlog/wineventlog_test.go

```diff
@@ -0,0 +1,163 @@
+	assert.Equal(t, NAME, elog.name)
+	assert.Equal(t, uint64(0), elog.eventOffset)
+	assert.Zero(t, elog.eventHandle)
+}
+
+// TestOpen verifies Open() succeeds with valid inputs.
+// And fails with invalid inputs.
+func TestOpen(t *testing.T) {
+	// happy
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	assert.NotNil(t, elog.eventHandle)
+	assert.NoError(t, elog.Close())
+	// bad name
+	elog = NewEventLog("FakeBadElogName", LEVELS, GROUP_NAME, STREAM_NAME,
+		RENDER_FMT, DEST, STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Error(t, elog.Open())
+	assert.Zero(t, elog.eventHandle)
+	// bad LEVELS does not cause Open() to fail.
+	// bad wlog.eventOffset does not cause Open() to fail.
+}
+
+// TestReadGoodSource will verify we can read events written by a registered
+// event log source.
+func TestReadGoodSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	// Without a sleep we do not read any events!!!
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 10, "[Application] [ERROR] [777] [CWA_UnitTest111] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadBadSource will verify that we cannot read events written by an
+// unregistered event log source.
+func TestReadBadSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 0, "[Application] [ERROR] [888] [CWA_UnitTest222] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadWithBothSources will verify we can read events written by a
+// registered event log source, even if the batch contains events from an
+// unregistered source too.
+func TestReadWithBothSources(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 10, "[Application] [ERROR] [777] [CWA_UnitTest111] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// seekToEnd skips past all current events in the event log.
+func seekToEnd(t *testing.T, elog *windowsEventLog) {
+	// loop until we stop getting records.
+	numRetrieved := 0
+	for {
+		records := elog.read()
+		//assert.Equal(t, nil, err)
+		t.Logf("seekToEnd() current count %v", len(records))
+		numRetrieved += len(records)
+		if len(records) == 0 {
+			break
+		}
+	}
+	t.Logf("seekToEnd() total %v", numRetrieved)
```

Comment:
Question - do you think there's any added value for having an assertion/check for the number of records that get ingested when we seek to the end? I'm not opposed to leaving it as is, but if it is an easily derivable number we can check against, I think it could help prevent regression down the road.

Reply:
We cannot.
We want the test to run on many different hosts.
We do not know the past history of events written to each hosts event log.
And even during our test we cannot guarantee that our test is the ONLY thing writing new events.
Which is why we get all events and just see if it at least contains the events our test wrote.